Secure File Removal Improvements

jentfoo opened the pull request:
This PR improves the reliability of secure file removal (particularly under conditions of unexpected errors), and leverages this functionality when removing certificates on disk.  This will have little value for SSD's but can be an effective measure when magnetic hard drives are used.

This PR fixes https://github.com/gravitational/teleport-private/issues/670

jentfoo committed:
utils.RemoveSecure: Still attempt a removal after error in overwrite

As extra caution, even if an error occurs during the overwrite process, we still want to attempt a removal of sensitive files.

jentfoo committed:
keystore.go: More secure removal of keyfiles

This commit ensures that deleted keyfiles have been overwritten.  This has little value on SSD's but can improve the security when the disk is magnetic.

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,15 +243,67 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 	return fileLock.Unlock, nil
 }
 
+// RemoveAllSecure is similar to `os.RemoveAll` however will delegate to the `RemoveSecure` implementation below
```
```suggestion
// RemoveAllSecure is similar to `os.RemoveAll` but leverages [RemoveSecure] to delete files.
```

- the square brackets create a link
- godoc is often viewed outside the source, so avoid saying "below"

jentfoo replied:
Updated, thank you!

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,15 +243,67 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	if info, err := os.Lstat(path); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		} else {
+			return trace.ConvertSystemError(err)
+		}
```
```suggestion
		}
		return trace.ConvertSystemError(err)
```

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,15 +243,67 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+		if len(removeErrors) > 0 {
+			if len(removeErrors) == 1 {
+				return trace.ConvertSystemError(removeErrors[0])
+			} else {
+				return fmt.Errorf("multiple errors in directory removal: %v", removeErrors)
+			}
```
```suggestion
			}
			return fmt.Errorf("multiple errors in directory removal: %v", removeErrors)

```

Also note, the use of errorf with %v will break `erorrs.Is` detection. Given that you're doing your best to preserve the error semantics of `os.RemoveAll`, we probably don't want this.

jentfoo replied:
`else` conditions were updated.  FWIW I feel like it improves readability to include the else when it's part of a chain of conditionals / thought process.  But I don't feel strongly about it (you just may have to keep catching me as this is very ingrained at this point)

zmb3 replied:
No worries. Will do. `golint` used to flag these..

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,15 +243,67 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+			}
+		}
+		return nil
+	} else { // file or symlink
+		return RemoveSecure(path)
+	}
```
```suggestion
	}
	// file or symlink
	return RemoveSecure(path)
```

General theme throughout this PR - you rarely need `else` in idiomatic Go code.

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,15 +243,67 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	} else if removeErr == nil {
+		return trace.ConvertSystemError(overwriteErr)
+	} else {
+		return trace.Errorf("multiple errors on removal: %v - %v", overwriteErr, removeErr)
```
Use `trace.NewAggregate` here instead?

Can get rid of the final else here too.

jentfoo replied:
Updated, thank you!

zmb3 reviewed (commented):
Haven't tried to run this yet, just did a pass of the code.

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,15 +242,64 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	} else if removeErr == nil {
+		return trace.ConvertSystemError(overwriteErr)
+	}
+	return trace.NewAggregate(overwriteErr, removeErr)
```
I wonder if we should even bother with an aggregate here. We really only care about the remove error, right? If the overwrite failed but we were still able to remove the file, there's not really any more action that the caller can take to handle the error - the file is gone.

jentfoo replied:
I considered this as well.  I wrote it as is because I felt like this was the most complete, but I had considered simply ignoring any overwrite error entirely.  Ultimately if the remove occurred but the overwrite failed for some reason, there is little actionable recourse the user can do.

Let me know your preference, it's fairly corner case either way and I don't feel strongly about it (beyond that I think it's important to try to remove regardless of potential overwrite errors)

zmb3 replied:
Don't feel strongly either way and agree it's probably a corner case.

That said, this is primarily intended to be used in user-facing client tools like tsh, and aggregate errors are pretty tricky to understand since they just concatenate two seemingly unrelated error messages. Given that, if I had to choose, I would lean towards just ignoring the overwrite error.

jentfoo replied:
Updated, thank you again!

zmb3 reviewed (approved):
Looks better, one last thought.

jentfoo commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,51 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+				removeErrors = append(removeErrors, err)
+			}
+		}
+	}
```
Good catch, fixed!

espadolini reviewed (commented):
Can we unlink the file first, and then overwrite its contents? As it is, anything opening the file while it still exists on disk but is in the process of being overwritten will just get some random garbage.

espadolini commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 	for i := 0; i < 3; i++ {
-		if err := overwriteFile(filePath); err != nil {
-			return trace.Wrap(err)
-		}
+		// ignore errors, ensure that os.Remove is attempted as that's the most critical action
+		_ = overwriteFile(filePath)
 	}
```
Can we just overwrite the file once with zeroes? Isn't that enough for any spinny storage made in the last 30 years or so?

jentfoo commented:
> Can we unlink the file first, and then overwrite its contents? As it is, anything opening the file while it still exists on disk but is in the process of being overwritten will just get some random garbage.

We can, but this would be an OS conditional logic.  I don't think in Windows you can write to files which have been removed from disk.  It would be only Linux and MacOS which would allow us to open the handle, remove, and then overwrite.

Your recommendation is superior when it can be supported, but since we would have to support both methods is it worth doubling the code?

AntonAM commented:
@jentfoo I don't think we have to double the code, there are already `fs_unix.go` and `fs_windows.go`, which can be used to provide platform-specific implementations. For example, you could add something like `unlinkFile` function in those, that we call before overwriting. And on mac/linux it would unlink the file, but on windows it would be no-op. Or something like that.

jentfoo commented:
@AntonAM Perhaps you see something I don't.  I don't believe you can simply no-op that single action unless you have other actions which are no-op in the unix case.  There is a need for action reordering with regards to the unlink / remove.

Unix:
1) Open
2) Unlink
3) Overwrite

Windows:
1) Overwrite
2) Remove

If we try to combine these we have a total steps of:
1) Open
2) Unlink (optional)
3) Overwrite
4) Remove (if not already unlinked)

While we could build the combined operation above, and that would technically be less lines.  But I feel like it's more confusing than a single conditional which chooses the flow that makes sense for the OS.

jentfoo commented:
I would be concerned about renaming, that would introduce a new place where the chain of operations may fail and result in a sensitive file being retained on disk.  Instead I would lean towards identify and improve any logic that could be impacted by this small overwrite window.

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+		}
+		return trace.ConvertSystemError(err)
+	} else if !info.IsDir() {
+		// file or symlink
```
Is there a point in doing RemoveSecure on a symlink? Should we just delete it as normal?

jentfoo replied:
I agree there is no value in a symlink case.  I documented it to be complete in the comment, but did not think it was a case that likely needed special handling.  Do you think it's worth the added code to handle symlinks differently?

codingllama replied:
Your call, but lets test with symlinks regardless.

jentfoo replied:
Added a test, it works as expected so I suggest we just keep the code minimal

codingllama replied:
Good, thanks!

jentfoo replied:
CI actually found a bug when the symlink points to a missing path.  So I may have to do explicit handling, looking now

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	}
+	// it's possible for a partial file list to be returned even if an error above was returned
+	for _, f := range files {
+		if err := RemoveAllSecure(filepath.Join(path, f.Name())); err != nil {
```
Is the name join necessary? Is that because we want an absolute path? If yes, it could use a comment or a named variable.

jentfoo replied:
How would you do it without the `Join`?  Sorry, the alternative is not obvious to me.

codingllama replied:
You're right, resolving.

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	var removeErrors []error
+	files, err := os.ReadDir(path)
+	if err != nil {
+		// don't fail fast, allow removal at end to be attempted
```
nit:

```suggestion
		// Don't fail fast, allow removal at end to be attempted.
```

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+		// don't fail fast, allow removal at end to be attempted
+		removeErrors = []error{err}
+	}
+	// it's possible for a partial file list to be returned even if an error above was returned
```
```suggestion
	// It's possible for a partial file list to be returned even if an error above was returned.
```

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+		return &os.PathError{Op: "RemoveAllSecure", Path: path, Err: syscall.EINVAL} // error type matches os.RemoveAll
+	}
+
+	if info, err := os.Lstat(path); err != nil {
```
FYI in case you want to try it, but there may be an elegant traversal solution using filepath.WalkDir. No need to change anything though.

https://pkg.go.dev/path/filepath#WalkDir

codingllama reviewed (commented):
Thanks for yet another security improvement, Jent!

AntonAM commented:
@jentfoo no, just wanted to highlight this option, I don't insist on the change :)

jentfoo commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 }
 
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
+func RemoveAllSecure(path string) error {
```
We can, though I did not see a reason that it would be valuable to do here.  Do you think this is a likely condition to hit?

codingllama replied:
It's a general purpose function, it pays to be careful. Particularly if we aim to replace stdlib use.

Let's make sure that the scenario is covered by tests, regardless of any special handling.

jentfoo replied:
Added and tested

jentfoo commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,51 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	} else if !info.IsDir() {
+		// file or symlink
+		return RemoveSecure(path)
+	}
```
Done

jentfoo commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
+func RemoveAllSecure(path string) error {
+	// Match os.RemoveAll protections in not permitting removal of "." directories
+	if path == "." || (len(path) >= 2 && path[len(path)-1] == '.' && os.IsPathSeparator(path[len(path)-2])) {
```
I added a comment, I don't think it warrants a whole function

jentfoo commented:
@AntonAM and @espadolini I updated it with the OS conditional logic, it ended up being cleaner than I had imagined

jentfoo commented on lib/utils/fs.go:
```diff
@@ -241,13 +242,53 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 }
 
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
+func RemoveAllSecure(path string) error {
```
This is covered by existing tests in `keystore`, but I added a specific test for this function in addition.

codingllama replied:
I don't think relying on indirect testing is good enough for an exported utility function. Let's make sure we cover the relevant corner cases, keystore is not concerned about that.

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,32 +242,82 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+// followed by calling os.Remove(filePath).
+func RemoveSecure(filePath string) error {
+	f, openErr := os.OpenFile(filePath, os.O_WRONLY, 0)
+	if os.IsNotExist(openErr) {
+		return openErr
+	} else if openErr == nil {
+		defer func() {
+			if closeErr := f.Close(); closeErr != nil {
+				log.WithError(closeErr).Warningf("Failed to close deleted file %v.", f.Name())
+			}
+		}()
 	}
```
Suggestion: cut nesting, explicitly handle `err!=nil`, drop Close() logging.

```suggestion
	switch {
	case os.IsNotExist(openErr):
		return trace. ConvertSystemError(openErr)
	case err != nil:
		// Attempt delete anyway.
		return trace.ConvertSystemError(os.Remove(filePath))
	}
	defer f.Close()
```

I don't think Close() errors are useful or interesting, but feel free to ignore that part.

codingllama commented on lib/utils/fs.go:
```diff
@@ -241,32 +242,82 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
-				log.WithError(closeErr).Warningf("Failed to close %v.", f.Name())
+
+	if runtime.GOOS == "windows" {
+		// windows can't unlink the file before overwriting
+		if f != nil {
+			for i := 0; i < 3; i++ {
+				_ = overwriteFile(f)
 			}
 		}
-	}()
+		return trace.ConvertSystemError(os.Remove(filePath))
+	} else {
+		removeErr := os.Remove(filePath)
+		if f != nil {
+			for i := 0; i < 3; i++ {
+				_ = overwriteFile(f)
+			}
+		}
+		return trace.ConvertSystemError(removeErr)
+	}
+}
```
Suggestion: drop separate branches. Builds on the previous suggestions by assuming `f` can't be nil in this block.

```suggestion
	isWindows := runtime.GOOS == "windows"
	var removeErr error
	if !isWindows {
		// Unlink first on unix systems.
		removeErr = os.Remove(filePath)
	}
	
	for i := 0; i < 3; i++ {
		_ = overwriteFile(f)
	}

	if isWindows {
		removeErr = os.Remove(filePath)
	}
		
	return trace.ConvertSystemError(removeErr)
}
```

jentfoo replied:
Mixing the conditional multiple times reduces readability to me, but I don't feel strongly about it

zmb3 replied:
Both look fine to me.

codingllama commented on lib/utils/fs.go:
```diff
@@ -280,6 +331,9 @@ func overwriteFile(filePath string) (err error) {
 
 	_, err = io.CopyN(f, rand.Reader, size)
+
+	f.Sync() // sync to ensure commit to hardware
```
Explicitly ignore error.

```suggestion
	_ = f.Sync() // sync to ensure commit to hardware
```

codingllama commented on lib/utils/fs_test.go:
```diff
@@ -347,3 +351,12 @@ func TestRemoveSecure(t *testing.T) {
 	require.True(t, os.IsNotExist(err), "File should be removed")
 }
+
+func TestRemoveSecureSymlink(t *testing.T) {
```
Suggestion: distinguish scenario from function being tested.

```suggestion
func TestRemoveSecure_symlink(t *testing.T) {
```

codingllama commented:
Looks good, mostly stylistic comments left.

codingllama commented on lib/utils/fs.go:
```diff
@@ -291,6 +291,18 @@ func RemoveAllSecure(path string) error {
+	if err != nil && os.IsNotExist(err) {
+		return nil
+	}
+	// don't fast return on other errors, still allow removeSecure to attempt removal
```
```suggestion
	// Don't fast return on other errors, still allow removeSecure to attempt removal.
```

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

jentfoo replied:
Happy to include this change.  However be aware there are thousands of comments in `teleport` that are not sentences (on a cursory search I found over 7000).  If this is a style we are adopting we should build a linter that enforces it.

codingllama replied:
A linter might be too disruptive given the state of the codebase, but I do think the advice is worth following. Not all comments need to be sentences, but when they are there's no downside in following it, plus it's a good habit to build, regardless of the programming language.

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,32 +241,97 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 	return fileLock.Unlock, nil
 }
 
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
```
I think we should also mention what makes it secure. Without that context, I might assume that the stdlib is plenty secure as compared to code that we wrote ourselves.

jentfoo replied:
Updated the docs, thank you!

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,32 +241,97 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
+	files, err := os.ReadDir(path)
+	if err != nil {
+		// Don't fail fast, allow removal at end to be attempted.
+		removeErrors = []error{err}
```
```suggestion
		removeErrors = append(removeErrors, err)
```

Append to the slice we already have rather than overwrite it with a new one?

zmb3 commented on lib/utils/fs.go:
```diff
@@ -241,32 +241,97 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
-				log.WithError(closeErr).Warningf("Failed to close %v.", f.Name())
+	defer f.Close()
+
+	if runtime.GOOS == "windows" {
```
nit: we do have a `WindowsOS` constant somewhere. Not sure if it's worth using, as it may just bloat the import graph.

zmb3 commented on lib/utils/fs.go:
```diff
@@ -280,6 +345,9 @@ func overwriteFile(filePath string) (err error) {
 
 	_, err = io.CopyN(f, rand.Reader, size)
+
+	_ = f.Sync() // sync to ensure commit to hardware
```
Should we be ignoring an error here?

Writes are buffered in many cases, meaning the error that we are checking (from `CopyN`) can be nil and the real error would surface here, but we're ignoring it.

zmb3 commented on lib/utils/fs_test.go:
```diff
@@ -316,16 +316,47 @@ func TestOverwriteFile(t *testing.T) {
+	require.True(t, os.IsNotExist(err), "File should be removed")
+}
+
+func TestRemoveSecure_symlink(t *testing.T) {
```
```suggestion
func TestRemoveSecureSymlink(t *testing.T) {
```

Not everyone agrees with me, but `_` are pretty uncommon in function names. I would remove it.

zmb3 reviewed (approved):
Approved, but I do think we need to check the sync error (see comment)

jentfoo commented on lib/utils/fs_test.go:
```diff
@@ -316,16 +316,47 @@ func TestOverwriteFile(t *testing.T) {
+	require.NoError(t, RemoveAllSecure(""))
+	require.NoError(t, RemoveAllSecure(tempDir))
+	_, err = os.Stat(tempDir)
+	require.True(t, os.IsNotExist(err), "Directory should be removed")
```
Done, thank you!

jentfoo committed:
Apply PR feedback, notably better testing and early unlinking if possible

This adds an OS conditional so that if possible the file will be removed and then overwritten using the previous file handle.
This will reduces the chance that the file will be witnessed with unexpected contents.
